### PR TITLE
Elasticsearch setup

### PR DESCRIPTION
**Pull request by rosemcc**
hub-search-proxy lambda functions:

- search - proxy for ElasticsSearch search queries
- update - upsert Contentful entry by id to ElasticSearch index. Contentful webhook endpoint.
- delete - delete Contentful entry by id from ElasticSearch index. Contentful webhook endpoint.
- bulk - endpoint for triggering export of searchable entries from Contentful, processing them and uploading them into the ElasticSearch index.

Also made changes to JenkinsFile for CI/CD of hub-search-proxy.

### CODE REVIEW

File: hub-search-proxy/handler.js [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+        let matchCount = 0;
+        for (let i = 0; i < queryOperators.length; i++) {
+          if (s.indexOf(queryOperators[i]) > -1) {
+           matchCount++;
+          }
+        };
+        return (matchCount > 0 || s.length < 3) ? s : s + fuzziness; 
+      }).join(' ');
+      
+      console.log(`Formatted query: ${formattedQueryString}`);
```
Comment (warrickct) [inline hub-search-proxy/handler.js:191]: Console log. Ignore this if intentional.

File: hub-search-proxy/handler.js [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+
+    const params = {
+      index: ELASTICSEARCH_INDEX_NAME,
+      body: query,
+      size: size,
+      from: from
+    }
+    
+    const result = await esClient.search(params);
+    console.log(`Found ${result.body.hits.total.value} results.`);
```
Comment (warrickct) [inline hub-search-proxy/handler.js:256]: Console log. Ignore this if intentional.

File: hub-search-proxy/handler.js [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    body: {
+      doc: doc,
+      doc_as_upsert: true  // if doc doesn't exist, create it
+    },
+    refresh: 'true'   // index refresh
+  };
+
+  try {
+    const result = await esClient.update(params);
+    console.log(`Processed document id ${result.body._id}: ${result.body.result}.`);
```
Comment (warrickct) [inline hub-search-proxy/handler.js:291]: Console log. Ignore this if intentional.

File: hub-search-proxy/handler.js [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+module.exports.delete = async (event, context) => {
+  const params = {
+    id: event.pathParameters.id,
+    index: ELASTICSEARCH_INDEX_NAME,
+    refresh: 'true'   // index refresh
+  };
+
+  try {
+    const result = await esClient.delete(params);
+    console.log(`Processed document id ${result.body._id}: ${result.body.result}.`);
```
Comment (warrickct) [inline hub-search-proxy/handler.js:317]: Console log. Ignore this if intentional.

File: hub-search-proxy/handler.js [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+    );
+  }
+}
+
+/**
+ * Bulk upload handler.
+ * Uses client helper: https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/7.x/client-helpers.html
+ */
+module.exports.bulk = async () => {
+  console.log("Bulk operation initiated.");
```
Comment (warrickct) [inline hub-search-proxy/handler.js:339]: Console log. Ignore this if intentional.

File: hub-search-proxy/handler.js [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+  
+  let validEntries;
+  const validContentTypes = ['article','caseStudy','equipment','event','service','software','subHub'];
+
+  // contentful export and filter entries
+  try {
+    console.log('Exporting data from Contentful space id: ' + spaceId);
```
Comment (warrickct) [inline hub-search-proxy/handler.js:346]: Console log. Ignore this if intentional.

File: hub-search-proxy/handler.js [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+      contentOnly: true,
+      downloadAssets: false,
+      saveFile: false
+    };
+    const contentfulData = await contentfulExport(options);
+    validEntries = contentfulData.entries.filter(
+      entry => validContentTypes.includes(entry.sys.contentType.sys.id)
+    );
+
+    console.log(`Found ${validEntries.length} entries to upload.`);
```
Comment (warrickct) [inline hub-search-proxy/handler.js:359]: Console log. Ignore this if intentional.

File: hub-search-proxy/handler.js [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+    const params = {
+      datasource: validEntries,
+      onDocument (doc) {
+        return [
+          { update: { _index: ELASTICSEARCH_INDEX_NAME, _id: doc.sys.id } },
+          { doc: doc, doc_as_upsert: true }
+        ]
+      },
+      onDrop (doc) {  // called everytime a document can’t be indexed and it has reached the maximum amount of retries.
+        console.log(doc);
```
Comment (warrickct) [inline hub-search-proxy/handler.js:371]: Console log. Ignore this if intentional.

File: hub-search-proxy/handler.js [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+      },
+      refreshOnCompletion: true, // Refresh the index after this
+      wait: 3000,   // How much time to wait before retries in milliseconds
+      retries: 3,   // How many times a document will be retried before to call the onDrop callback
+      concurrency: 5,  // How many request will be executed at the same time.
+      flushBytes: 5000000,  // The size of the bulk body in bytes to reach before to send it. Default of 5MB.
+    };
+
+    // perform the upload
+    console.log(`Uploading documents to index: ${ELASTICSEARCH_INDEX_NAME}`);
```
Comment (warrickct) [inline hub-search-proxy/handler.js:381]: Console log. Ignore this if intentional.

File: hub-search-proxy/test/mock_query_basic.json [hunk 10]
```diff
@@ -0,0 +1,9 @@
+{
+    "event": {
+        "httpMethod": "POST",
+        "headers": { "Content-Type": "application/json" },
+        "body": {
+            "comment": "Tis better to have tested and lost than never to have tested at all"
```
Comment (warrickct) [inline hub-search-proxy/test/mock_query_basic.json:6]: Tis true

**Review by warrickct — APPROVED** [review]
From what I can see it looks alright to me. There's console logs in the handler code but if they're in there for CloudWatch logs then I think that's fine.

**Comment by rosemcc** [comment]
Thanks @warrickct I removed a couple of console logs and added a few more changes. Tests passing in sandbox

File: hub-search-proxy/test/mock_query_basic.json (hunk 10 above)
Comment (Hganavak) [reply hub-search-proxy/test/mock_query_basic.json:6]: Thanks @warrickct , I thought I was pretty philosophical here #im14andthisisdeep

Comment (rosemcc) [reply hub-search-proxy/test/mock_query_basic.json:6]: @Hganavak thank you Sam, gave me a chuckle when I first ran the test

**Pull request merged**